sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricResultsMatchers.java

```diff
@@ -43,9 +43,50 @@
     return metricsResult(namespace, name, step, value, true);
   }
 
+  /**
+   * Matches a {@link MetricResult} with the given namespace, name and step, and a matcher for the
+   * value for either committed or attempted (based on {@code isCommitted}) metrics.
+   */
+  public static <T> Matcher<MetricResult<T>> metricsResult(
+      final String namespace,
+      final String name,
+      final String step,
+      final Matcher<T> valueMatcher,
+      final boolean isCommitted) {
+
+    final String metricState = isCommitted ? "committed" : "attempted";
+    return new MatchNameAndKey<T>(namespace, name, step) {
+      @Override
+      protected boolean matchesSafely(MetricResult<T> item) {
+        final T metricValue = isCommitted ? item.getCommitted() : item.getAttempted();
+        return super.matchesSafely(item) && valueMatcher.matches(metricValue);
+      }
+
+      @Override
+      public void describeTo(Description description) {
+        super.describeTo(description);
+        description.appendText(String.format(", %s=", metricState));
+        valueMatcher.describeTo(description);
+        description.appendText("}");
+      }
+
+      @Override
+      protected void describeMismatchSafely(MetricResult<T> item, Description mismatchDescription) {
+        final T metricValue = isCommitted ? item.getCommitted() : item.getAttempted();
+        super.describeMismatchSafely(item, mismatchDescription);
+        mismatchDescription.appendText(String.format("%s: ", metricState));
+        valueMatcher.describeMismatch(metricValue, mismatchDescription);
+        mismatchDescription.appendText("}");
+      }
+    };
+  }
+
   /**
    * Matches a {@link MetricResult} with the given namespace, name and step, and whose value equals
    * the given value for either committed or attempted (based on {@code isCommitted}) metrics.
+   *
+   * <p>For metrics with a {@link {@link GaugeResult}}, only the value is matched and the timestamp
+   * is ignored.
    */
   public static <T> Matcher<MetricResult<T>> metricsResult(
```

Comment:
Another note: I checked that these modifications to existing methods produce the same error messages as before by manually forcing other existing metrics tests to fail.

Reply:
If I understand this correctly, you did diff of failing tests output before and after the change and there was no difference?

Reply:
Exactly: I modified some tests to fail and checked the output before and after the change.

*For exact match:*

```
Expected: a collection containing MetricResult{inNamespace="source", name="elements_read", step="GenerateSequence/Read(UnboundedCountingSource)", attempted=<999L>}
     but: mismatches were: [MetricResult{attempted: <999L> != <1000L>}]
```

*Using a matcher (new):*

```
Expected: a collection containing MetricResult{inNamespace="source", name="elements_read", step="GenerateSequence/Read(UnboundedCountingSource)", attempted=a value equal to or greater than <1000000L>}
     but: mismatches were: [MetricResult{attempted: <1000L> was less than <1000000L>}]
```